File: SSUMobile/Modules/About/Views/SSUFeedbackViewController.m
//
//  SSUFeedbackViewController.m
//  SSUMobile
//
//  Created by Eric Amorde on 11/27/14.
//  Copyright (c) 2014 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUFeedbackViewController.h"
#import "SSULogging.h"
#import "SSUMoonlightCommunicator.h"
#import <MBProgressHUD/MBProgressHUD.h>

static NSString * SSUUserDefaultsLastFeedbackDate = @"LastFeedbackDate";
static NSTimeInterval WAIT_PERIOD = 60.0;

@interface SSUFeedbackViewController ()

@property (nonatomic) BOOL hasBegunEditing;
@property (nonatomic) BOOL submitting;
@property (nonatomic,strong) NSString * placeholderText;
@property (nonatomic,strong) UIBarButtonItem * sendButton;

@end

@implementation SSUFeedbackViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.hasBegunEditing = NO;
    self.placeholderText = self.textView.text;
    
    self.sendButton = [[UIBarButtonItem alloc] initWithTitle:@"Send" style:UIBarButtonItemStyleDone target:self action:@selector(sendButtonPressed)];
    self.navigationItem.rightBarButtonItem = self.sendButton;
    self.sendButton.enabled = NO;
    
    [[NSUserDefaults standardUserDefaults] registerDefaults:@{
                                                              SSUUserDefaultsLastFeedbackDate : @([[NSDate distantPast] timeIntervalSinceReferenceDate])
                                                              }];
}

#pragma mark - UITextViewDelegate

- (void) textViewDidBeginEditing:(UITextView *)textView {
    if (!self.hasBegunEditing) {
        // The first time the user begins editing, clear the placeholder text.
        textView.text = @"";
        self.hasBegunEditing = YES;
    }
}

- (void) textViewDidChange:(UITextView *)textView {
    NSString * text = [textView.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    self.sendButton.enabled = ![text isEqualToString:@""];
}

- (void) textViewDidEndEditing:(UITextView *)textView {
    NSString * text = [textView.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if ([text isEqualToString:@""]) {
        // If the user didn't enter anything, reset to the original state
        textView.text = self.placeholderText;
        self.hasBegunEditing = NO;
    }
}

#pragma mark - Button Actions

- (void) sendButtonPressed {
    NSString * content = [self.textView.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    NSString * email = [self.emailTextField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if ([content isEqualToString:@""]) {
        return;
    }
    [self sendFeedbackWithContent:content email:email];
    
}

#pragma mark - Other

/**
 Send feedback to moonlight
 
 @param content The text of the feedback/suggestion
 @param email The email of the user, or an empty string
 */
- (void) sendFeedbackWithContent:(NSString *)content email:(NSString *)email {
    if (self.submitting) {
        return;
    }
    if (![self canSubmitFeedback]) {
        [self alertWithTitle:@"Unable to Send feedback"
                     message:@"You recently submitted some feedback to us. Please wait a few minutes before sending another submission"];
        return;
    }
    self.submitting = YES;
    NSString * baseURL = [SSUMoonlightBaseURL stringByAppendingPathComponent:@"createFeedback"];
    NSString * version = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleShortVersionString"];
    id params = @{
                  @"content" : content,
                  @"email" : email,
                  @"version" : version,
                  };
    [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    [SSUMoonlightCommunicator postURL:[NSURL URLWithString:baseURL]
                           parameters:params
                           completion:^(NSURLResponse * response, NSData * data, NSError * error) {
                         [MBProgressHUD hideAllHUDsForView:self.view animated:YES];
                         self.submitting = NO;
                         if (error) {
                             [self alertWithTitle:@"Error" message:@"Sorry, something went wrong. Try again later."];
                             SSULogError(@"Error during feedback submission: %@",error);
                         }
                         else {
                             [self alertWithTitle:@"Success" message:@"Thank you for your feedback!"];
                             self.emailTextField.text = @"";
                             self.textView.text = @"";
                             [[NSUserDefaults standardUserDefaults] setDouble:[[NSDate date] timeIntervalSinceReferenceDate] forKey:SSUUserDefaultsLastFeedbackDate];
                         }
                     }];
}

/**
 Only allow the user to submit multiple entries after waiting for a
 defined time period
 */
- (BOOL) canSubmitFeedback {
    NSTimeInterval lastSubmission = [[NSUserDefaults standardUserDefaults] doubleForKey:SSUUserDefaultsLastFeedbackDate];
    NSDate * date = [NSDate dateWithTimeIntervalSinceReferenceDate:lastSubmission];
    NSTimeInterval sinceLastSubmission = ABS([date timeIntervalSinceNow]);
    return sinceLastSubmission >= WAIT_PERIOD;
}

- (void) alertWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertView * alert = [[UIAlertView alloc] initWithTitle:title
                                                     message:message
                                                    delegate:nil
                                           cancelButtonTitle:@"OK"
                                           otherButtonTitles:nil];
    [alert show];
}

@end

File: SSUMobile/Modules/About/Views/SSUFeedbackViewController.swift
//
//  SSUFeedbackViewController.swift
//  SSUMobile
//
//  Created by Eric Amorde on 3/25/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation

class SSUFeedbackViewController: UITableViewController, UITextViewDelegate {
    
    @IBOutlet var textView: UITextView!
    @IBOutlet var emailTextField: UITextField!
    
    private var hasBegunEditing = false
    private var isSubmitting = false
    private var placeholderText: String?
    private var sendButton: UIBarButtonItem = UIBarButtonItem(title: "Send", style: .done, target: self, action: #selector(SSUFeedbackViewController.sendButtonPressed))
    
    override func viewDidLoad() {
        super.viewDidLoad()
        placeholderText = textView.text // set in storyboard
        
        navigationItem.rightBarButtonItem = sendButton
        sendButton.isEnabled = false
    }
    
    // MARK: UITextViewDelegate
    
    func textViewDidBeginEditing(_ textView: UITextView) {
        if !hasBegunEditing {
            // The first time the user begins editing, clear the placeholder text.
            textView.text = ""
            hasBegunEditing = true
        }
    }
    
    func textViewDidChange(_ textView: UITextView) {
        let text = textView.text.trimmingCharacters(in: .whitespacesAndNewlines)
        sendButton.isEnabled = !text.isEmpty
    }
    
    func textViewDidEndEditing(_ textView: UITextView) {
        let text = textView.text.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.isEmpty {
            // If the user ended editing without entering text, put the placeholder back
            textView.text = placeholderText
        }
    }
    
    // MARK: Button actions
    
    @objc
    private func sendButtonPressed() {
        let content = textView.text.trimmingCharacters(in: .whitespacesAndNewlines)
        let email = emailTextField.text?.trimmingCharacters(in: .whitespacesAndNewlines)
        if !content.isEmpty {
            sendFeedbackSubmission(content: content, email: email)
        }
    }
    
    // MARK: Send submission
    
    /**
     Send feedback to moonlight
     
     - parameter content The text of the feedback/suggestion
     - parameter email The email of the user, or an empty string
     */
    private func sendFeedbackSubmission(content: String, email: String?) {
        if !SSUAboutModule.instance.canSubmitFeedback {
            alert(title: "Unable to submit feedback", message: "You recently submitted some feedback to us. Please wait a few minutes before sending another submission")
            return
        }
        isSubmitting = true
        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        let osVersion = ProcessInfo.processInfo.operatingSystemVersionString
        let params: [String:Any] = [
            "os_name": "ios",
            "content": content,
            "email": email ?? "",
            "app_version": appVersion ?? "Unknown",
            "os_version": osVersion
        ]
        let hud = MBProgressHUD.showAdded(to: tableView, animated: true)
        SSUMoonlightCommunicator.postPath("ssumobile/feedback/", parameters: params) { (response, data, error) in
            if error != nil {
                self.alert(title: "Error", message: "There was an error when submitting your feedback. Please try again or come back later")
            } else {
                DispatchQueue.main.async {
                    hud.hide(animated: true)
                    SSUConfiguration.sharedInstance().lastFeedbackDate = Date()
                    self.alert(title: "Success", message: "Thank you for your feeedback!")
                    self.emailTextField.text = nil
                    self.textView.text = self.placeholderText ?? self.textView.text
                }
            }
        }
    }
    
    private func alert(title: String, message: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        DispatchQueue.main.async {
            self.present(alert, animated: true, completion: nil)
        }
    }
}
